### PR TITLE
LUCENE-9831: Hunspell GeneratingSuggester: faster flag & case checks, less allocations

### PR DESCRIPTION
**Pull request by donnerpeter**
# Description

Yet another attempt to slim down the CPU usage for each enumerated word

# Solution

Mostly it's avoiding excessive method calls and moving allocations as late as possible

# Tests

No new tests, ~5-10% speedup in `TestPerformance.de_suggest`

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java [hunk 1]
```diff
@@ -60,9 +63,15 @@
       String word, WordCase originalCase) {
     Comparator<Weighted<Root<String>>> natural = Comparator.naturalOrder();
     PriorityQueue<Weighted<Root<String>>> roots = new PriorityQueue<>(natural.reversed());
-    List<Root<String>> entries = new ArrayList<>();
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:63]: We don't allocate Root now until we're sure they're suitable, and then we add them directly to the PriorityQueue above

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java [hunk 2]
```diff
@@ -60,9 +63,15 @@
...
+    EntryFilter filter = new EntryFilter(dictionary);
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:66]: and to check for suitability, we create a special object

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java [hunk 3]
```diff
@@ -60,9 +63,15 @@
...
     boolean ignoreTitleCaseRoots = originalCase == WordCase.LOWER && !dictionary.hasLanguage("de");
-    TrigramAutomaton automaton = new TrigramAutomaton(word);
+    TrigramAutomaton automaton =
+        new TrigramAutomaton(word) {
+          @Override
+          char transformChar(char c) {
+            return dictionary.caseFold(c);
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:72]: We don't lowercase the entire string before calculating ngrams anymore, we lowercase each character as we process it

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java [hunk 4]
```diff
@@ -75,25 +84,29 @@
...
-          String root = rootChars.toString();
-          filterSuitableEntries(root, forms, entries);
-          if (entries.isEmpty()) return;
+          int suitable = filter.findSuitableFormIndex(forms, 0);
+          if (suitable < 0) return;
 
-          if (ignoreTitleCaseRoots && WordCase.caseOf(rootChars) == WordCase.TITLE) {
+          if (ignoreTitleCaseRoots
+              && Character.isUpperCase(rootChars.charAt(0))
+              && WordCase.caseOf(rootChars) == WordCase.TITLE) {
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:92]: `caseOf` walks the entire array, but we don't always need that, e.g. if the first letter is lowercase (the majority of entries)

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java [hunk 5]
```diff
@@ -75,25 +84,29 @@
...
+              automaton.ngramScore(rootChars)
+                  - longerWorsePenalty(word.length(), rootChars.length)
+                  + commonPrefix(word, rootChars);
 
           if (roots.size() == MAX_ROOTS && sc < roots.peek().score) {
             return;
           }
 
-          entries.forEach(e -> roots.add(new Weighted<>(e, sc)));
+          String root = rootChars.toString();
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:105]: Only after all checks do we allocate the string and `Root` objects

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FlagEnumerator.java [hunk 6]
```diff
@@ -79,6 +79,36 @@ boolean hasFlag(int entryId, char flag) {
       return false;
     }
 
+    boolean hasAnyFlag(int entryId, char[] sortedFlags) {
```
[the remark this answers is not in the document]
Comment (rmuir) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FlagEnumerator.java:82]: I don't have a problem with the changes here, but it makes me wonder if a bitset is more appropriate data structure? sorry if its a dumb idea, i dont remember the properties of all these flags. but If we are assigning flags numeric ids already, can we get away with using a single bit? bitset intersection has the advantage of using vector instructions.

[the remark this answers is not in the document]
Comment (donnerpeter) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FlagEnumerator.java:82]: There can be a lot of flags (up to 64K), so bit sets would be pretty large

**Pull request merged**